### ISSUE TITLE
'reportUnnecessaryIsInstance'  can lead to missed refactoring on exhaustive narrowing of unions

### ISSUE THREAD

**Issue by Tomaz-Vieira** [microsoft/pyright#6147]
**Describe the bug**
Say that one has a function that takes a Union and, for every type in that Union,  returns a value that does not depend on the actual value of that Union:

```python
# pyright: strict

from typing import Union
from typing_extensions import TypeAlias

MyUnion: TypeAlias = Union[int, str]

def decide_on_union(u: MyUnion) -> int:
    if isinstance(u, int):
        return 1
    if isinstance(u, str):
        return 2
```

The last `if isinstance` is marked by pyright as unnecessary, which incentivizes the user to just remove it:

```python
def decide_on_union(u: MyUnion) -> int:
    if isinstance(u, int):
        return 1
    return 2
```

If in the future that user adds a new variant to the union, like:

```python
MyUnion: TypeAlias = Union[int, str, None]
``` 

then they will not be warned that `decide_on_union` is now incorrectly treating two or more variants in the same way, so it seems that this warning can actually be damaging.

The current workaround is to leave the last `isinstance`, ignore the warnings by adding `#pyright: ignore [reportUnnecessaryIsInstance]`, and then be reminded that `MyUnion` got a new variant when pyright warns that the ignore itself is no longer necessary.

Could we have a different heuristic for this error? Maybe just not apply it when one is narrowing Unions?

**VS Code extension or command-line**
extension version 1.1.331

**Comment by minmax** [comment]
If you use pattern matching, pyright will show error when new type will bee added to union.

**Comment by erictraut** [comment]
This is working as designed. The `reportUnnecessaryComparison` check is not enabled by default in standard type checking. If you decide to enable it, you are opting in to its behavior. Adding special-case handling for your specific use case would be hard to justify.

If you want to keep this diagnostic check enabled, there are ways to change your code to avoid it in this case. Or if you don't want to change your code, you can use a `# type: ignore` or `# pyright: ignore` to suppress the error.

**Comment by Ivanca** [comment]
Use `# pyright: ignore[reportUnnecessaryIsInstance]` instead of `# pyright: ignore` so any other potential failed checks do trigger a warning
